.github/workflows/ci.yml

```diff
@@ -1,30 +1,131 @@
----
-
 name: CI
 
 on:
   pull_request:
-    branches: [devel]
+    branches:
+      - devel
 
   push:
-    branches: [devel]
+    branches:
+      - devel
 
 jobs:
+  sanity:
+    name: ${{ matrix.test.name }}
+    runs-on: ubuntu-20.04
+    container:
+      image: quay.io/ansible/ansible-builder-test-container:1.0.0
+      env:
+        PIP_CACHE_DIR: ${{ runner.temp }}/.cache/pip
+
+    strategy:
+      fail-fast: false
+      matrix:
+        test:
+          - name: Lint
+            env: linters
+
+          - name: Docs
+            env: docs
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Setup test environment
+        run: tox -e ${{ matrix.test.env }} --notest
+
+      - name: Run tests
+        run: tox -e ${{ matrix.test.env }}
+
+      - name: Save docs
+        uses: actions/upload-artifact@v2
+        if: ${{ success() && matrix.test.name == 'Docs' }}
+        with:
+          name: docs
+          path: docs/build/html
+          retention-days: 3
+
   integration:
     runs-on: ubuntu-20.04
-    name: Integration
+    name: Integration - ${{ matrix.py_version.name }}
+    needs: images
+
+    strategy:
+      fail-fast: false
+      matrix:
+        py_version:
+          - name: 3.8
+            tox: py38
+
+          - name: 3.9
+            tox: py39
+
     steps:
-      - uses: actions/checkout@v2
+      - name: Checkout
+        uses: actions/checkout@v2
 
-      - uses: actions/setup-python@v2
+      - name: Install Python ${{ matrix.py_version.name }}
+        uses: actions/setup-python@v2
         with:
-          python-version: "3.8"
+          python-version: ${{ matrix.py_version.name }}
 
-      - name: Install dependencies
+      - name: Setup test environment
         run: |
-          python -m pip install --upgrade pip
-          pip install tox
+          ${{ format('python{0}', matrix.py_version.name) }} -m pip install --upgrade pip
+          ${{ format('python{0}', matrix.py_version.name) }} -m pip install tox
```

Comment:
I wouldn't put this many separate commands in one step because it's hard to parse where one starts and the other one ends in the logs. Unless you add runner commands for collapsing the log groups around each.